tools/approval-service/cmd/approval-service.go

```diff
@@ -114,43 +134,73 @@ func NewGitHubEventSource(processor ApprovalProcessor) *GitHubEventSource {
 // Setup GH client, webhook secret, etc.
 // https://github.com/go-playground/webhooks may help here
 func (ghes *GitHubEventSource) Setup() error {
-	// TODO
+	deployReviewChan := make(chan *github.DeploymentReviewEvent)
+	ghes.deployReviewChan = deployReviewChan
+
+	mux := http.NewServeMux()
+	eventProcessor := webhook.EventHandlerFunc(func(event interface{}) error {
+		switch event := event.(type) {
+		case *github.DeploymentReviewEvent:
+			deployReviewChan <- event
+			return nil
+		default:
+			return trace.Errorf("unknown event type: %T", event)
+		}
+	})
+	mux.Handle("/webhook", webhook.NewHandler(
+		eventProcessor,
+		webhook.WithSecretToken([]byte("secret-token")), // TODO: get from config
+		webhook.WithLogger(logger),
+	))
+
+	ghes.srv = &http.Server{
+		Addr:    ghes.addr,
+		Handler: mux,
+	}
+
 	return nil
 }
 
 // Take incoming events and respond to them
-func (ghes *GitHubEventSource) Run(done chan struct{}) error {
-	// If anything errors, deny the request. For safety, maybe `defer`
-	// the "response" function?
-	go func() {
-		// Notify the service that the listener is completely done.
-		// Normally this should only be hit if there is a fatal error
-		defer func() { done <- struct{}{} }()
+func (ghes *GitHubEventSource) Run(ctx context.Context) error {
+	errc := make(chan error)
 
-		// Incoming webhook payloads
-		// This should be closed by the webhook listener func
-		payloads := make(chan interface{})
-
-		// Listen for webhook calls
-		go ghes.listenForPayloads(payloads, done)
+	// Start the HTTP server
+	go func() {
+		logger.Info("Listening for GitHub Webhooks", "address", ghes.addr)
+		errc <- ghes.srv.ListenAndServe()
+		close(errc)
+	}()
 
-		for payload := range payloads {
-			go ghes.processWebhookPayload(payload, done)
+	// Process incoming events
+	go func() {
+		defer close(ghes.deployReviewChan)
+		logger.Info("Starting GitHub event processor")
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case deployReview := <-ghes.deployReviewChan:
+				// Process the event
+				go ghes.processDeploymentReviewEvent(deployReview)
+			}
 		}
 	}()
 
-	return nil
-}
-
-// Listen for incoming webhook events. Register HTTP routes, start server, etc. Long running, blocking.
-func (ghes *GitHubEventSource) listenForPayloads(payloads chan interface{}, done chan struct{}) {
-	// Once a call is received, it should return a 200 response immediately.
+	var err error
+	// This will block until an error occurs or the context is done
+	select {
+	case err = <-errc:
+		ghes.srv.Shutdown(context.Background()) // Ignore error - we're already handling one
+	case <-ctx.Done():
+		err = ghes.srv.Shutdown(context.Background())
```

Comment:
These should probably take a time-limited shutdown context stemming from `context.Background()`.